src/stability_sdk/animation.py

```diff
@@ -231,167 +229,144 @@ def setup_animation(self):
+            ))
         elif args.animation_mode == '3D':
-            op = warp3d_op(
-                    border_mode = args.border,
-                    translate_x = frame_args.translation_x_series[frame_idx],
-                    translate_y = frame_args.translation_y_series[frame_idx],
-                    translate_z = frame_args.translation_z_series[frame_idx],
-                    rotate_x = frame_args.rotation_x_series[frame_idx],
-                    rotate_y = frame_args.rotation_y_series[frame_idx],
-                    rotate_z = frame_args.rotation_z_series[frame_idx],
-                    near_plane = args.near_plane,
-                    far_plane = args.far_plane,
-                    fov = frame_args.fov_series[frame_idx],
-                )
-            ops.append(op)
+            ops.append(warp3d_op(
+                border_mode = args.border,
+                translate_x = frame_args.translation_x_series[frame_idx],
+                translate_y = frame_args.translation_y_series[frame_idx],
+                translate_z = frame_args.translation_z_series[frame_idx],
+                rotate_x = frame_args.rotation_x_series[frame_idx],
+                rotate_y = frame_args.rotation_y_series[frame_idx],
+                rotate_z = frame_args.rotation_z_series[frame_idx],
+                near_plane = args.near_plane,
+                far_plane = args.far_plane,
+                fov = frame_args.fov_series[frame_idx],
+            ))
 
         elif args.animation_mode == 'Video Input':
-            video_extract_nth = args.extract_nth_frame
-            for i in range(video_extract_nth):
+            for _ in range(args.extract_nth_frame):
                 success, video_next_frame = self.video_reader.read()
             if success:
-                video_next_frame = cv2.resize(video_next_frame, (args.W, args.H), interpolation=cv2.INTER_LANCZOS4)
+                video_next_frame = cv2.resize(
+                    video_next_frame, 
+                    (args.width, args.height), 
+                    interpolation=cv2.INTER_LANCZOS4
+                )
                 if args.video_flow_warp:
-                    op = warpflow_op(
+                    ops.append(warpflow_op(
                         prev_frame=self.video_prev_frame,
                         next_frame=video_next_frame,
-                    )
-                    ops.append(op)
+                    ))
                 self.video_prev_frame = video_next_frame
-                color_match_image = video_next_frame
+                self.color_match_image = video_next_frame
 
-        return ops, color_match_image
+        return ops
 
-    def render_animation(
-        self,
-        stub,
-        args=None,
-        out_dir=None
-    ):
-
-        if not args:
-            args = self.args
-        if not out_dir:
-            out_dir = self.out_dir
-        key_frame_values = self.keyframe_values
+    def render(self) -> Generator[Image.Image, None, None]:
```

Comment:
this is the opposite direction I want to take this function. removing the ability to pass in arguments makes this harder to both test and use in isolation. I ultimately want the different components of the animation process here to become less coupled rather than more tightly coupled. I'm planning to refactor this further to give the user the ability to intervene at for-loop iteration granularity. we can include push-button run functionality like this `render()` implementation, but I want to make sure we have the lower level constructs as well (which `render` would then wrap). while this is all we've got, I want to expose more of its internals rather than obfuscate them

Reply:
oh the problem I saw with passing args again in here is that it can break the animation. setup_animation prepares the internal state based on the mode and parameters and if those change at the call to render it'll bug out. better would be methods to make changes to whatever state you want to mutate and that can ensure it doesn't break internal state and assumptions. render is called once for each frame so caller could request some frames, then make changes to, then request more frames no problem.

Reply:
regarding `setup_animation` specifically, I was thinking pushing that up and out of this class entirely. so the presumption would be that users would parameterize their animation into a form analogous to what `setup_animation` currently outputs, and then feed that into the animator as the argument parameterization. this has the added benefit to making the workhorse functionality agnostic to DSL choice, which will make it easier to plug DS in as a backend to stuff that isn't compliant with deforum's notation